File: frontend/src/api/websites.ts
import apiClient from './client'

export interface Website {
  id: string
  project_id: string
  url: string
  title?: string
  description?: string
  status: 'pending' | 'crawling' | 'analyzing' | 'completed' | 'failed'
  crawled_at?: string
  analyzed_at?: string
  digital_twin?: Record<string, any>
  created_at: string
  updated_at: string
}

export interface CreateWebsiteRequest {
  project_id: string
  url: string
}

export interface UpdateWebsiteRequest {
  title?: string
  description?: string
  status?: string
}

export const websitesApi = {
  list: async (projectId?: string, page = 1, pageSize = 20) => {
    const params = new URLSearchParams({
      page: String(page),
      page_size: String(pageSize),
      ...(projectId && { project_id: projectId }),
    })
    const response = await apiClient.get(`/api/v1/websites?${params}`)
    return response.data
  },

  create: async (data: CreateWebsiteRequest): Promise<Website> => {
    const response = await apiClient.post('/api/v1/websites', data)
    return response.data
  },

  get: async (websiteId: string): Promise<Website> => {
    const response = await apiClient.get(`/api/v1/websites/${websiteId}`)
    return response.data
  },

  update: async (websiteId: string, data: UpdateWebsiteRequest): Promise<Website> => {
    const response = await apiClient.patch(`/api/v1/websites/${websiteId}`, data)
    return response.data
  },

  delete: async (websiteId: string): Promise<void> => {
    await apiClient.delete(`/api/v1/websites/${websiteId}`)
  },

  analyze: async (websiteId: string, maxPages = 50, depth = 3, includeAssets = true) => {
    const response = await apiClient.post(`/api/v1/websites/${websiteId}/analyze`, {
      max_pages: maxPages,
      depth,
      include_assets: includeAssets,
    })
    return response.data
  },

  getAnalysisStatus: async (websiteId: string) => {
    const response = await apiClient.get(`/api/v1/websites/${websiteId}/analyze/status`)
    return response.data
  },

  getPages: async (websiteId: string, page = 1, pageSize = 20) => {
    const params = new URLSearchParams({
      page: String(page),
      page_size: String(pageSize),
    })
    const response = await apiClient.get(`/api/v1/websites/${websiteId}/pages?${params}`)
    return response.data
  },

  getAssets: async (websiteId: string, assetType?: string, page = 1, pageSize = 20) => {
    const params = new URLSearchParams({
      page: String(page),
      page_size: String(pageSize),
      ...(assetType && { asset_type: assetType }),
    })
    const response = await apiClient.get(`/api/v1/websites/${websiteId}/assets?${params}`)
    return response.data
  },

  getDigitalTwin: async (websiteId: string) => {
    const response = await apiClient.get(`/api/v1/websites/${websiteId}/digital-twin`)
    return response.data
  },
}

File: backend/app/api/websites.py
"""Website management API endpoints."""

from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status, Query

from app.dependencies import (
    get_current_user,
    get_website_service,
    get_current_website,
)
from app.models.user import User
from app.models.website import Website
from app.services.website_service import WebsiteService
from app.schemas.website import (
    WebsiteCreate,
    WebsiteUpdate,
    WebsiteResponse,
    WebsiteListResponse,
    WebsiteAnalyzeRequest,
)
from app.schemas.common import PaginatedResponse
from app.core.errors import NotFoundError, ValidationError

router = APIRouter()


@router.get("/", response_model=PaginatedResponse[WebsiteListResponse])
async def list_websites(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    project_id: Optional[UUID] = None,
    current_user: User = Depends(get_current_user),
    website_service: WebsiteService = Depends(get_website_service),
):
    """List websites for the current user."""
    skip = (page - 1) * page_size
    
    websites, total = await website_service.list_websites(
        user_id=current_user.id,
        project_id=project_id,
        skip=skip,
        limit=page_size,
    )
    
    return PaginatedResponse(
        items=websites,
        total=total,
        page=page,
        page_size=page_size,
        pages=(total + page_size - 1) // page_size,
    )


@router.post("/", response_model=WebsiteResponse, status_code=status.HTTP_201_CREATED)
async def create_website(
    request: WebsiteCreate,
    current_user: User = Depends(get_current_user),
    website_service: WebsiteService = Depends(get_website_service),
):
    """Add a website to a project."""
    try:
        website = await website_service.create_website(
            project_id=request.project_id,
            url=request.url,
            user_id=current_user.id,
        )
        return website
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )


@router.get("/{website_id}", response_model=WebsiteResponse)
async def get_website(
    website: Website = Depends(get_current_website),
):
    """Get website by ID."""
    return website


@router.patch("/{website_id}", response_model=WebsiteResponse)
async def update_website(
    request: WebsiteUpdate,
    website: Website = Depends(get_current_website),
    website_service: WebsiteService = Depends(get_website_service),
):
    """Update website."""
    updated = await website_service.update_website(
        website_id=website.id,
        updates=request.dict(exclude_unset=True),
    )
    return updated


@router.delete("/{website_id}")
async def delete_website(
    website: Website = Depends(get_current_website),
    website_service: WebsiteService = Depends(get_website_service),
):
    """Delete website."""
    await website_service.delete_website(website.id)
    return {"message": "Website deleted successfully"}


@router.post("/{website_id}/analyze")
async def analyze_website(
    request: WebsiteAnalyzeRequest,
    website: Website = Depends(get_current_website),
    website_service: WebsiteService = Depends(get_website_service),
):
    """Analyze website content."""
    job_id = await website_service.analyze_website(
        website_id=website.id,
        max_pages=request.max_pages,
        depth=request.depth,
        include_assets=request.include_assets,
    )
    return {
        "message": "Analysis started",
        "job_id": job_id,
        "website_id": website.id,
    }


@router.get("/{website_id}/analyze/status")
async def get_analysis_status(
    website: Website = Depends(get_current_website),
    website_service: WebsiteService = Depends(get_website_service),
):
    """Get website analysis status."""
    status = await website_service.get_analysis_status(website.id)
    return status


@router.get("/{website_id}/pages")
async def list_website_pages(
    website: Website = Depends(get_current_website),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    website_service: WebsiteService = Depends(get_website_service),
):
    """List pages crawled from a website."""
    skip = (page - 1) * page_size
    pages, total = await website_service.list_pages(
        website_id=website.id,
        skip=skip,
        limit=page_size,
    )
    
    return {
        "items": pages,
        "total": total,
        "page": page,
        "page_size": page_size,
        "pages": (total + page_size - 1) // page_size,
    }


@router.get("/{website_id}/assets")
async def list_website_assets(
    website: Website = Depends(get_current_website),
    asset_type: Optional[str] = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    website_service: WebsiteService = Depends(get_website_service),
):
    """List assets extracted from a website."""
    skip = (page - 1) * page_size
    assets, total = await website_service.list_assets(
        website_id=website.id,
        asset_type=asset_type,
        skip=skip,
        limit=page_size,
    )
    
    return {
        "items": assets,
        "total": total,
        "page": page,
        "page_size": page_size,
        "pages": (total + page_size - 1) // page_size,
    }


@router.get("/{website_id}/digital-twin")
async def get_digital_twin(
    website: Website = Depends(get_current_website),
    website_service: WebsiteService = Depends(get_website_service),
):
    """Get website digital twin."""
    twin = await website_service.get_digital_twin(website.id)
    if not twin:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Digital twin not found. Please analyze the website first.",
        )
    return twin